[DOCUMENTATION][feature/otx] Explanation section. PR 2 (#1566)
open-edge-platform/training_extensions: docs/source/explanation/algorithms/segmentation/instance_segmentation.rst

Instance Segmentation
=====================

| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.

> | It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows for more detailed information about the object, such as its shape and location, to be extracted.

Reviewer: ```suggestion
| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows for more detailed information about the object, such as its shape and location, to be extracted.
```
```suggestion
| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
```

> | Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.

Reviewer: Double enter after this line to start a new paragraph. I'd also suggest using image for visualization like [this](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/instance-segmentation-security-0002/assets/instance-segmentation-security-0002.png)

For the supervised training we use the following algorithms components:

.. _instance_segmentation_supervised_pipeline:

- ``Augmentations``: We use only a random flip for both augmentations pipelines, train and validation.

- ``Optimizer``: We use `SGD <https://en.wikipedia.org/wiki/Stochastic_gradient_descent>`_ optimizer with the weight decay set to **1e-4** and momentum set to **0.9**.

- ``Learning rate schedule``: For scheduling training process we use **ReduceLrOnPlataeu** with linear learning rate warmup for **200** iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.

- ``Loss functions``: For the bounding box regression we use **L1 Loss** (the sum of the absolute differences between the ground truth value and the predicted value), `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_ for the categories classification and segmentation masks prediction.

> - Additionally, we use  **Exponential Moving Average (EMA)** for the model's weights and **early stopping** to add adaptability to the training pipeline and prevent overfitting.

Reviewer: ```suggestion
- Additionally, we use  the **Exponential Moving Average (EMA)** for the model's weights and the **early stopping** to add adaptability to the training pipeline and prevent overfitting.
```

+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+
| `Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50 <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/detection/configs/instance_segmentation/resnet50_maskrcnn/template.yaml>`_                | MaskRCNN-ResNet50          | 533.80              | 177.90          |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+

> ``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and Flops and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade off beetween accuracy and speed. It is a better choice when training time and computational cost are in priority.

Reviewer: ```suggestion
``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and FLOPs and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade-off between accuracy and speed. It is a better choice when training time and computational cost are in priority.
```

Instance Segmentation
=====================

| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.

> | It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.

Reviewer: ```suggestion
It is a more advanced version of object detection, as it doesn't only detect the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
```

> | Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.

Reviewer: ```suggestion
Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.
```

.. _instance_segmentation_image_example:

.. image:: ../../../../utils/images/instance_seg_example.png
  :width: 600

| We solve this problem in the `MaskRCNN <https://arxiv.org/abs/1703.06870>`_ manner. The main idea of Mask R-CNN is to add a branch for predicting an object mask in parallel with the existing branch for bounding box regression and object classification.
| This is done by using a fully convolutional network (FCN) on top of the feature map generated by the last convolutional layer of the backbone network. The model first generates region proposals, and then uses a RoIAlign layer to align the region proposals with the feature map, then the FCN predicts the class and box offset for each proposal and the mask for each class.

> | The mask branch is trained to predict a binary mask for each object instance, where the mask is aligned with the object's bounding box and has the same size as the region of interest (RoI). The predicted mask is then used to segment the object from the background.

Reviewer: ```suggestion
We solve this problem in the `MaskRCNN <https://arxiv.org/abs/1703.06870>`_ manner. The main idea of Mask R-CNN is to add a branch for predicting an object mask in parallel with the existing branch for bounding box regression and object classification.
This is done by using a fully convolutional network (FCN) on top of the feature map generated by the last convolutional layer of the backbone network. The model first generates region proposals, and then uses a RoIAlign layer to align the region proposals with the feature map, then the FCN predicts the class and box offset for each proposal and the mask for each class.
The mask branch is trained to predict a binary mask for each object instance, where the mask is aligned with the object's bounding box and has the same size as the region of interest (RoI). The predicted mask is then used to segment the object from the background.
```

For the supervised training we use the following algorithms components:

.. _instance_segmentation_supervised_pipeline:

- ``Augmentations``: We use only a random flip for both augmentations pipelines, train and validation.

- ``Optimizer``: We use `SGD <https://en.wikipedia.org/wiki/Stochastic_gradient_descent>`_ optimizer with the weight decay set to **1e-4** and momentum set to **0.9**.

> - ``Learning rate schedule``: For scheduling training process we use **ReduceLrOnPlataeu** with linear learning rate warmup for **200** iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.

Reviewer: This `patience` supposed to be highlighted? Because it's rendered in other way
![image](https://user-images.githubusercontent.com/33455576/216109134-d3e095b0-6116-4850-9cf3-3a825d54c6bf.png)

Author: No, it is basically brackets, but I will remove them to not mislead anyone

- ``Loss functions``: For the bounding box regression we use **L1 Loss** (the sum of the absolute differences between the ground truth value and the predicted value), `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_ for the categories classification and segmentation masks prediction.

- Additionally, we use the **Exponential Moving Average (EMA)** for the model's weights and the **early stopping** to add adaptability to the training pipeline and prevent overfitting.

**************
Dataset Format
**************

> For the dataset handling inside OTX, we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. For instance segmentation we support `COCO <https://cocodataset.org/#format-data>` and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ dataset formats.

Reviewer: ```suggestion
For the dataset handling inside OTX, we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. For instance segmentation we support `COCO <https://cocodataset.org/#format-data>`_ and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ dataset formats.
```

If you have your dataset in those formats, then you can simply run OTX using one line of code:

.. code-block::

    $ otx {train, optimize} <model_template> --train-data-root <path_to_data_root>

.. note::

    OTX will recognize train and validation datasets automatically. If you want to emphasize the validation dataset root path, you can simply add the ``--val-data-root <path_to_val_data_root>`` option.

>     If you have your annotations in COCO format, but, for example, have the structure or names differ from the original one - you can directly pass paths to annotation files to the command line with two CLI options: ``--train-ann-files <path_to_train_annotations.json>``, ``--val-ann-files <path_to_val_annotations.json>``.

Reviewer: Also, shouldn't the same note be in the `Semantic Segmentation` page?

>     Please, refer to our :doc:`dedicated tutorial <../../../tutorials/base/how_to_train/instance_segmentation>` how to train, validate and optimize instance segmentation model for more details.

Reviewer: Add additional enter before this paragraph to make a new point.

******
Models
******

We support the following ready-to-use model templates:

+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+
| Template ID                                                                                                                                                                                                                                | Name                       | Complexity (GFLOPs) | Model size (MB) |
+============================================================================================================================================================================================================================================+============================+=====================+=================+

> | `Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/detection/configs/instance_segmentation/efficientnetb2b_maskrcnn/template.yaml>`_  | MaskRCNN-EfficientNetB2B   | 68.48               | 13.27           |

Reviewer 2: Could you update links to `develop`?
```suggestion
| `Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B <https://github.com/openvinotoolkit/training_extensions/blob/develop/otx/algorithms/detection/configs/instance_segmentation/efficientnetb2b_maskrcnn/template.yaml>`_  | MaskRCNN-EfficientNetB2B   | 68.48               | 13.27           |
```

| `Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50 <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/detection/configs/instance_segmentation/resnet50_maskrcnn/template.yaml>`_                | MaskRCNN-ResNet50          | 533.80              | 177.90          |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+

``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and FLOPs and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade-off between accuracy and speed. It is a better choice when training time and computational cost are in priority.

> In the table below the `mAP <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ metric on some academic datasets using our :ref:`supervised pipeline <instance_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 1024x1024 image resolution, for other hyperparameters, please, refer to the related template. We trained all models on 1 GPU Nvidia GeForce GTX3090.

Reviewer 2: ```suggestion
In the table below the `mAP <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ metric on some academic datasets using our :ref:`supervised pipeline <instance_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 1024x1024 image resolution, for other hyperparameters, please, refer to the related template. We trained each model with single Nvidia GeForce RTX3090.
```

Author: agree, this sounds  better
